src/sc_stage.py

```diff
@@ -49,129 +51,140 @@ def __init__(self, node_process, time_per_batch=3):
+
         self.node = node_process.node
         self.send_message = node_process.sendMessage  # function to send messages to cluster
 
         self.inbound_log = FileDict(abspath("./tmp/" + self.name + ".inbound.log"))
         self.outbound_log = FileDict(abspath("./tmp/" + self.name + ".outbound.log"))
 
-        # dict to keep track of requests that haven't gotten
-        # a response and need additional retries.
+        # dict to keep track of requests that haven't gotten a response yet
+        # maps request_id -> Message object
         self.pending_requests = {}
 
         # hack to stop stage
         self.running = Event()
         self.running.set()  # set the stage to run by default
 
         self.manufacture_count = 0
+        self.consumed_count = 0
         self._attempt_log_recovery()
 
+        log.info("Node %d's stage bootstrap complete", self.node_id)
+
     def _attempt_log_recovery(self):
         '''
             attempts to populate the in-memory data structures from
-            data in the disk log if there is any.
-
-            uses raftos
-
-            The log is unique to the stage using the stage-name as the key.
+            data in the WAL.
+        '''
+        out_log, in_log = self.outbound_log, self.inbound_log
+        if not out_log.size() and not in_log.size():
+            log.info("Node %d's batch status WALs are empty. Skipping post-crash item and material status verification", self.node_id)
+            return
 
+        flow = self.state_helper.get_flow()
+        self.metrics.increase_metric(self.node_id, "flow_queries")
+        while not flow:
+            # cant proceed until flow is obtained
+            sleep(self.time_per_batch)
+            log.debug("Node %d waiting for flow to be set to start batch status verification", self.node_id)
+            flow = self.state_helper.get_flow()
+            self.metrics.increase_metric(self.node_id, "flow_queries")
+
+        self._attempt_inbound_recovery(in_log, flow)
+        self._attempt_outbound_recovery(out_log, flow)
+
+    def _attempt_inbound_recovery(self, in_log, flow):
+        log.info("Node %d attempting post-crash recovery to verify status of batches in inbound WAL of size %d", self.node_id, in_log.size())
+
+        neighbors = [nid for nid, _ in flow.getIncomingFlowsForNode(self.node_id)]
+        unknown_state_batches = set()
+
+        for batch, state in in_log.items():
+            # only consumed batches need no further confirmation from
+            # neighbors about the status of the item
+            if state != BatchStatus.CONSUMED:
+                self._request_batch_status_from_neighbor(neighbors, batch)
+                unknown_state_batches.add(batch)
+            else:
+                self.consumed_count += 1
+
+        for batch in unknown_state_batches:
+            self.outbound_log[batch] = BatchStatus.WAITING_FOR_NEIGHBOR
+
+    def _attempt_outbound_recovery(self, out_log, flow):
```

Comment:
Ah missed this earlier. We don't need outbound recovery, right? As discussed, it's the responsibility of the consumer to send an ack. Whenever we receive the ack, we just update the state. Need to remove this.

Reply:
good point. Will remove outbound recovery if it's not needed. Thought 2-way recovery will be helpful as a safety measure.Let me think about it.